File: src/python/caption.py
import ika
from thing import Thing

class Caption(Thing):

    def __init__(self, engineRef, font, text, x = None, y = None, duration=200):
        self.engineRef = engineRef
        self.font = font

        width = self.font.StringWidth(text)
        height = self.font.height

        if x is None:   self.x = (self.engineRef.video.xres - width) // 2
        else:           self.x = x

        if y is None:   self.y = self.engineRef.video.yres - height - 40
        else:           self.y = y

        self.text = text
        self.opacity = 0
        self.duration = duration
        self._updateGen =  self._update()
        self.update = lambda: next(self._updateGen)

    def _update(self):
        while self.opacity < 256:
            self.opacity += 2
            yield None

        while self.duration > 0:
            self.duration -= 1
            yield None

        while self.opacity > 0:
            self.opacity -= 2
            yield None

        yield True # seppuku

    def draw(self):
        o = min(255, self.opacity)
        self.font.PrintWithOpacity(self.x, self.y, self.text, o)

File: src/winter/caption.ts
import { Thing } from "./thing.js"
import { Engine, FontClass, PyEngine } from "./winter.js"

export class Caption extends Thing {
    private x: number
    private y: number
    private opacity: number
    private updateGen: IterableIterator<boolean>

    constructor(
        engineRef: PyEngine,
        private font: FontClass,
        private text: string,
        x: number | null = null,
        y: number | null = null,
        private duration: number = 200,
    ) {
        super()
        const engine: Engine = engineRef.getEngine()

        const width = this.font.StringWidth(text)
        const height = this.font.height

        this.x = (x !== null) ? x : Math.floor((engine.video.xres - width) / 2)
        this.y = (y !== null) ? y : engine.video.yres - height - 40

        this.opacity = 0
        this.updateGen =  this._update()
    }

    private *_update() {
        while (this.opacity < 256) {
            this.opacity += 2
            yield false
        }

        while (this.duration > 0) {
            this.duration -= 1
            yield false
        }

        while (this.opacity > 0) {
            this.opacity -= 2
            yield false
        }

        yield true // seppuku
    }

    update() {
        const { value } = this.updateGen.next()
        return value
    }

    draw() {
        const o = Math.min(255, this.opacity)
        this.font.PrintWithOpacity(this.x, this.y, this.text, o)
    }
}
